### PR TITLE
[MRG+2] add flags to request

### PR DESCRIPTION
**Pull request by foromer4**
Issue #2047, by pawelmhm

### CODE REVIEW

**Comment by kmike** [comment]
I think these flags will disappear when Requst is serialized to a disk queue - `scrapy.util.reqser` should be changed. Also, it looks like request.replace should keep flags attribute instead of discarding it.

There is a small efficiency penalty - e.g. with 1M requests in queue it is at least 64MB. Maybe it worths it to store empty flags as None internally, but I'm not sure (users should be able to do flags.append without checking for None). Opinions? Is efficiency penalty OK?

Another way to get more information out of Requests is to use custom Request subclasses which override `__repr__`; it doesn't work because of https://github.com/scrapy/scrapy/issues/1890.

File: scrapy/utils/reqser.py [hunk 1]
```diff
@@ -58,7 +59,8 @@ def request_from_dict(d, spider=None):
         meta=d['meta'],
         encoding=d['_encoding'],
         priority=d['priority'],
-        dont_filter=d['dont_filter'])
+        dont_filter=d['dont_filter'],
+        flags= d['flags'])
```
Comment (kmike) [inline scrapy/utils/reqser.py]: I think it should be d.get('flags'), for backwards compatibility, i.e. to be able to continue a long crawl after Scrapy update. 

Maybe it makes sense not to serialize flags if flags is None, but I'm not sure about that.

**Comment by kmike** [comment]
Hey @foromer4,

It looks better. I think docs are still needed: this feature should be mentioned here: https://github.com/scrapy/scrapy/blob/master/docs/topics/request-response.rst

**Comment by foromer4** [comment]
Hi, I think all requested changes are in, is it possible to merge into main branch?

File: scrapy/utils/reqser.py [hunk 2]
```diff
@@ -31,6 +31,7 @@ def request_to_dict(request, spider=None):
         '_encoding': request._encoding,
         'priority': request.priority,
         'dont_filter': request.dont_filter,
+        'flags' : request.flags
```
Comment (kmike) [inline scrapy/utils/reqser.py]: could you please remove a whitespace before `:`?

File: scrapy/utils/reqser.py [hunk 3]
```diff
@@ -58,7 +59,8 @@ def request_from_dict(d, spider=None):
         meta=d['meta'],
         encoding=d['_encoding'],
         priority=d['priority'],
-        dont_filter=d['dont_filter'])
+        dont_filter=d['dont_filter'],
+        flags= d.get('flags'))
```
Comment (kmike) [inline scrapy/utils/reqser.py]: there is an extra space after `=`

File: tests/test_utils_reqser.py [hunk 4]
```diff
@@ -25,7 +25,8 @@ def test_all_attributes(self):
             cookies={'currency': u'руб'},
             encoding='latin-1',
             priority=20,
-            meta={'a': 'b'})
+            meta={'a': 'b'},
+            flags= ['testFlag'])
```
Comment (kmike) [inline tests/test_utils_reqser.py]: There is an extra space after `=` - could you plese remove it? We're trying to follow PEP8 in Scrapy source code, with 80 columns limit somewhat relaxed.

**Comment by foromer4** [comment]
Hopefully documentation and spacing issues are now resolved.

File: tests/test_utils_reqser.py [hunk 5]
```diff
@@ -53,6 +54,7 @@ def _assert_same_request(self, r1, r2):
         self.assertEqual(r1._encoding, r2._encoding)
         self.assertEqual(r1.priority, r2.priority)
         self.assertEqual(r1.dont_filter, r2.dont_filter)
+        self.assertEquals(r1.flags, r2.flags)
```
Comment (kmike) [inline tests/test_utils_reqser.py]: According to https://docs.python.org/3.5/library/unittest.html#deprecated-aliases assertEquals is a deprecated alias to assertEqual - could you please use assertEqual, like in above lines?

Comment (foromer4) [inline tests/test_utils_reqser.py]: Right, fixed it now. thanks

Comment (foromer4) [inline tests/test_utils_reqser.py]: Are there any more issues I should fix?

Comment (foromer4) [inline tests/test_utils_reqser.py]: should I try to further advance this feature or discard it?

File: scrapy/http/request/__init__.py [hunk 6]
```diff
@@ -36,6 +36,7 @@ def __init__(self, url, callback=None, method='GET', headers=None, body=None,
         self.dont_filter = dont_filter
 
         self._meta = dict(meta) if meta else None
+        self.flags = None if flags is None else list(flags)
```
Comment (redapple) [inline scrapy/http/request/__init__.py]: Is this to address @kmike 's https://github.com/scrapy/scrapy/pull/2082#issuecomment-231215967 ?
I haven't tested but does "_users should be able to do flags.append without checking for None_ work"?

If `None` is to be kept, what about `self.flags = list(flags) if flags else None`?

File: tests/test_utils_reqser.py (hunk 5 above)
Comment (kmike) [inline tests/test_utils_reqser.py]: @foromer4 I like this feature, definitely +1 to finish it. Sorry for us being so picky!

File: scrapy/http/request/__init__.py (hunk 6 above)
Comment (foromer4) [inline scrapy/http/request/__init__.py]: Thanks , missed that one... now pushing a change

File: tests/test_utils_reqser.py (hunk 5 above)
Comment (foromer4) [inline tests/test_utils_reqser.py]: thanks for the patience ,  hope I addressed all the issues you raised please let me know if I missed anything

**Comment by kmike** [comment]
Looks good 👍 
@foromer4 could you please sqash the commits?

**Comment by foromer4** [comment]
squashed commits

**Comment by kmike** [comment]
@foromer4 something went wrong - there is a lot of unrelated commits in this PR. Could you please rebase it on master?

**Comment by foromer4** [comment]
Please excuse my inexperience...
I tried rebasing and than squashing using rebase HEAD~[X]

But I think it caused some mess. If I understand, you can squash during the merge process of the fork:
https://github.com/blog/2141-squash-your-commits

But I don't think I have the privileges for this.

**Comment by redapple** [comment]
Looks good to me. We can use GitHub's "Squash and merge" option here.

**Comment by kmike** [comment]
@redapple let's try it :)

**Comment by kmike** [comment]
Thanks @foromer4 for the persistence!

**Comment by foromer4** [comment]
Thank you, for the patience :)

**Pull request merged**